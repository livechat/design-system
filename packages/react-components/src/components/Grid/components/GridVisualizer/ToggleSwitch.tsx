import * as React from 'react';

import { Switch } from '../../../Switch';

import styles from './ToggleSwitch.module.scss';

interface ToggleSwitchProps {
  checked: boolean;
  onChange: (checked: boolean) => void;
  label: string;
}

export const ToggleSwitch: React.FC<ToggleSwitchProps> = ({
  checked,
  onChange,
  label,
}) => {
  return (
    <div className={styles.container}>
      <Switch
        on={checked}
        onChange={(_, value) => onChange(value)}
        ariaLabel={label}
        size="medium"
      />
      <span>{label}</span>
    </div>
  );
};
