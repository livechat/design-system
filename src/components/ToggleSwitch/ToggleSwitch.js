import * as React from 'react';
import * as PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styles from './style.scss';
import getMergedClassNames from '../../utils/getMergedClassNames';

const cx = classNames.bind(styles);
const acceptedSizes = ['basic', 'compact'];
const noop = () => {};

class ToggleSwitch extends React.PureComponent {
  static propTypes = {
    on: PropTypes.bool,
    defaultOn: PropTypes.bool,
    size: PropTypes.oneOf(acceptedSizes),
    onToggle: PropTypes.func
  };

  static defaultProps = {
    size: 'basic',
    onToggle: noop,
    defaultOn: false
  };

  state = {
    on: this.isControlledByProps() ? this.props.on : this.props.defaultOn,
    prevPropsOn: this.props.on
  };

  static getDerivedStateFromProps(props, state) {
    if (props.on !== state.prevPropsOn) {
      return {
        on: props.on,
        prevPropsOn: props.on
      };
    }
    return null;
  }

  toggleState = e => {
    const hasCb = this.props.onToggle !== noop;
    if (hasCb) {
      this.props.onToggle(e, this.state.on);
      return;
    }
    e.preventDefault();
    this.setState({ on: !this.state.on });
  };

  isControlledByProps() {
    return this.props.on !== undefined;
  }

  render() {
    const { className, size } = this.props;
    const { on } = this.state;
    const baseClass = 'toggleSwitch';

    const mergedClassNames = getMergedClassNames(
      cx({
        [baseClass]: true,
        [`${baseClass}--disabled`]: !on,
        [`${baseClass}--enabled`]: on,
        [`${baseClass}--${size}`]: acceptedSizes.some(s => s === size)
      }),
      className
    );

    return (
      <div onClick={e => this.toggleState(e)} className={mergedClassNames} />
    );
  }
}

export default ToggleSwitch;
