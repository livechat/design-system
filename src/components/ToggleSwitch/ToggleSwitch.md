Use toggles to immediately switch between two states (e.g. ON / OFF)

```js
<div style={{display: 'flex'}}>
  <span style={{marginRight: '10px', alignItems: 'center'}}> Enable sound notifications for customers </span>
  <ToggleSwitch />
</div>
```

<h3>Best practices</h3>
<ul>
  <li>Donâ€™t require users to press a button to confirm the change. </li>
  <li>If you need users to press a button, use checkboxes instead. </li>
  <li>Write clear labels to make the choice less ambigous.</li>
  <li>Use positive and active wording for labels </li>
</ul>