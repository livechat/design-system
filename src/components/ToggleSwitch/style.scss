@import '../../scss/variables/colors';


$base-class: 'toggleSwitch';

.#{$base-class} {
  border-radius: 34px;
  position: relative;
  &:before {
    position: absolute;
    content: "";
    bottom: 1px;
    background-color: #fff;
    border-radius: 50%;
  }
  &--disabled {
    background-color: #DDE2E5;
    &:before {
      left: 1px;
    }
  }
  &--enabled {
    background-color: $success-color;
    &:before {
      right: 1px;
    }
  }
  &--regular {
    width: 44px;
    height: 20px;
    &:before {
      height: 19px;
      width: 19px;
    }
  }

  &--compact {
    width: 26px;
    height: 17px;
    &:before {
      height: 15px;
      width: 15px;
    }
  }
}