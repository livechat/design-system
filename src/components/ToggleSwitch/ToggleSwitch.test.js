import * as React from 'react';
import { shallow } from 'enzyme';
import ToggleSwitch from './ToggleSwitch';

describe('ToggleSwitch', () => {
  it('should render basic ToggleSwitch', () => {
    const component = shallow(<ToggleSwitch />);

    expect(component).toMatchSnapshot();
  });

  it('should call on toggle when clicked', () => {
    const toggleFn = jest.fn();
    const component = shallow(<ToggleSwitch onToggle={toggleFn} />);
    component.simulate('click');
    expect(toggleFn).toHaveBeenCalled();
  });
});
