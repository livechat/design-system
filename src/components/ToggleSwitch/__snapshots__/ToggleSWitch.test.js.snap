// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch should render basic ToggleSwitch 1`] = `
<div
  className="toggleSwitch toggleSwitch--disabled toggleSwitch--basic"
  onClick={[Function]}
/>
`;
